monitor/cert_submitter.go

```diff
@@ -0,0 +1,143 @@
+	certStats *certSubmitterStats = &certSubmitterStats{
+		certSubmitLatency: promauto.NewHistogramVec(prometheus.HistogramOpts{
+			Name:    "cert_submit_latency",
+			Help:    "Latency submitting certificate chains to CT logs",
+			Buckets: internetFacingBuckets,
+		}, []string{"uri"}),
+		certSubmitResults: promauto.NewCounterVec(prometheus.CounterOpts{
+			Name: "cert_submit_results",
+			Help: "Count of results from submitting certificate chains to CT logs, sliced by status",
+		}, []string{"uri", "status"}),
+	}
+)
+
+// certSubmitter is a type for periodically issuing certificates and submitting
+// them to a log's add-chain endpoint.
+type certSubmitter struct {
+	logger *log.Logger
+	clk    clock.Clock
+	client monitorCTClient
+	logURI string
+	stats  *certSubmitterStats
+
+	// How long to sleep between submitting certificates to the log
+	certSubmitInterval time.Duration
+	// ECDSA private key used to issue certificates to submit to the log. Nil if
+	// no certificates are to be submitted to the log.
+	certIssuerKey *ecdsa.PrivateKey
+	// Certificate used as the issuer for certificates submitted to the log. Nil
+	// if no certificates are to be submitted to the log. The Certificate's public
+	// key must correspond to the private key in certIssuerKey.
+	certIssuer *x509.Certificate
+}
+
+// run starts a goroutine that calls submitCertificate, sleeps for
+// certSubmitInterval and then repeats forever.
+func (c *certSubmitter) run() {
+	go func() {
+		for {
+			c.submitCertificate()
+			c.logger.Printf("Sleeping for %s before next certificate submission\n",
+				c.certSubmitInterval)
+			c.clk.Sleep(c.certSubmitInterval)
+		}
+	}()
+}
+
+// submitCertificate issues a certificate with the certSubmitter's
+// certIssuer/certIssuerKey and submits it to a monitored log's add-chain
+// endpoint. The latency of the submission is tracked in the
+// `cert_submit_latency` prometheus histogram. If the submission fails, or the
+// returned SCT is invalid the `cert_submit_results` prometheus countervec is
+// incremented with a "fail" status tag. If the submission succeeds the
+// `cert_submit_results` prometheus countervec is incremented with a "ok" status
+// tag. An SCT is considered invalid if the signature does not validate, or if
+// the timestamp is too far in the future or the past (controlled by
+// `requiredSCTFreshness`).
+func (c *certSubmitter) submitCertificate() {
+	c.logger.Printf("Submitting certificate to %q\n", c.logURI)
+
+	cert, err := pki.IssueTestCertificate(c.certIssuerKey, c.certIssuer, c.clk)
+	if err != nil {
+		// This should not occur and if it does we should abort hard
+		panic(fmt.Sprintf("!!! Error issuing certificate: %s\n", err.Error()))
+	}
+
+	// Because ct-woodpecker issues directly off of a fake root the "chain" only
+	// contains the leaf certificate we minted in issueCertificate()
+	chain := []ct.ASN1Cert{
+		ct.ASN1Cert{Data: cert.Raw},
+	}
+
+	latencyLabels := prometheus.Labels{"uri": c.logURI}
```

Comment:
I think separating these label declarations from the sites where they are used makes it a little harder to read the code. Could you move these lines to right above where the variables are referenced for the first time?